internal/storage/inmem/store.go

```diff
@@ -0,0 +1,328 @@
+	idx, err := incrementEventIndex(tx)
+	if err != nil {
+		return nil
+	}
+	tx.Commit()
+
+	s.publishEvent(idx, &pbresource.WatchEvent{
+		Operation: pbresource.WatchEvent_OPERATION_DELETE,
+		Resource:  res,
+	})
+
+	return nil
+}
+
+// List resources of the given type, tenancy, and optionally matching the given
+// name prefix.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) List(typ storage.UnversionedType, ten *pbresource.Tenancy, namePrefix string) ([]*pbresource.Resource, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+
+	return listTxn(tx, query{typ, ten, namePrefix})
+}
+
+// WatchList watches resources of the given type, tenancy, and optionally
+// matching the given name prefix.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) WatchList(typ storage.UnversionedType, ten *pbresource.Tenancy, namePrefix string) (*Watch, error) {
+	// If the user specifies a wildcard, we subscribe to events for resources in
+	// all partitions, peers, and namespaces, and manually filter out irrelevant
+	// stuff (in Watch.Next).
+	//
+	// If the user gave exact tenancy values, we can subscribe to events for the
+	// relevant resources only, which is far more efficient.
+	var sub stream.Subject
+	if ten.Partition == storage.Wildcard ||
+		ten.PeerName == storage.Wildcard ||
+		ten.Namespace == storage.Wildcard {
+		sub = wildcardSubject{typ}
+	} else {
+		sub = tenancySubject{typ, ten}
+	}
+
+	ss, err := s.pub.Subscribe(&stream.SubscribeRequest{
+		Topic:   eventTopic,
+		Subject: sub,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Watch{
+		sub: ss,
+		query: query{
+			resourceType: typ,
+			tenancy:      ten,
+			namePrefix:   namePrefix,
+		},
+	}, nil
+}
+
+// OwnerReferences returns the IDs of resources owned by the resource with the
+// given ID.
+//
+// For more information, see the storage.Backend documentation.
+func (s *Store) OwnerReferences(id *pbresource.ID) ([]*pbresource.ID, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+
+	iter, err := tx.Get(tableNameResources, indexNameOwner, id)
+	if err != nil {
+		return nil, err
+	}
+
+	var refs []*pbresource.ID
+	for v := iter.Next(); v != nil; v = iter.Next() {
+		refs = append(refs, v.(*pbresource.Resource).Id)
+	}
+	return refs, nil
+}
+
+const (
+	tableNameMetadata  = "metadata"
+	tableNameResources = "resources"
+
+	indexNameID    = "id"
+	indexNameOwner = "owner"
+
+	metaKeyEventIndex = "index"
+)
+
+func listTxn(tx *memdb.Txn, q query) ([]*pbresource.Resource, error) {
+	iter, err := tx.Get(tableNameResources, indexNameID+"_prefix", q)
+	if err != nil {
+		return nil, err
+	}
+
+	list := make([]*pbresource.Resource, 0)
+	for v := iter.Next(); v != nil; v = iter.Next() {
+		res := v.(*pbresource.Resource)
+
+		if q.matches(res) {
+			list = append(list, res)
+		}
+	}
+	return list, nil
+}
+
+type meta struct {
+	Key   string
+	Value any
+}
+
+func incrementEventIndex(tx *memdb.Txn) (uint64, error) {
```

Comment:
Is the event index generally just a way of having a monotonically increasing value for all writes that wont be coupled with the Raft index?

Reply:
Yep, exactly. It's purely for the event publishing machinery (and splicing the snapshot together with subsequent events).